### DISCUSSION TITLE
how to set unload time for each vehicle and for each unload depot

### DISCUSSION THREAD

**Discussion by orangiee** [google/or-tools#2914] [Routing (and legacy CP) questions] [answered]
Hello, 
I'am working on your cvrp_reload.py to solve a routing problem. 

An objectif is to let all drivers come back to unload depot and rest for one hour and then go for another route. So my question is how to set this 'one hour rest at the depot' ? for example, driver A finish the first route at 11:50 and come back to unload depot at 12:15. then how can we set his departure time at 13:15 for his second route. and each driver can has his departure time which depends on his arrival time of his first route. Thank you in advance !

**Comment by Mizux** [comment, marked as answer]
simply add a "service time" of 1hour in the transit callback (or transit matrix) for any arc `unload_node -> *`

the `transit_callback(A, B)` must contains the `service_time(A)` + the `travel_time(A, B)`

**Reply by orangiee** [reply]
hi, thanks  a lot for your fast reply. When you say: add a ''service time'', do you mean that I create another 'service time' if node is in the list of unload depot, then service time == one hour; I'm not sure how to add this . Could you please tell me ? Thanks ! 
```python
def create_time_evaluator(data):
    """Creates callback to get total times between locations."""

    def service_time(data, node):
        """Gets the service time for the specified location."""
        return abs(data['demands'][node]) * data['time_per_demand_unit']

    def travel_time(data, from_node, to_node):
        """Gets the travel times between two locations."""
        if from_node == to_node:
            travel_time = 0
        else:
            travel_time = manhattan_distance(
                    data['locations'][from_node], data['locations'][to_node]) / data['vehicle_speed']
        return travel_time

    _total_time = {}
    # precompute total time to have time callback in O(1)
    for from_node in range(data['num_locations']):
        _total_time[from_node] = {}
        for to_node in range(data['num_locations']):
            if from_node == to_node:
                _total_time[from_node][to_node] = 0
            else:
                _total_time[from_node][to_node] = int(
                    service_time(data, from_node) + travel_time(
                        data, from_node, to_node))
```

**Reply by Mizux** [reply]
you could rewrite your `service_time()` function, something like this

```python
data['unload_node_list'] = {13, 31, 42, ....}

    def service_time(data, node):
        """Gets the service time for the specified location."""
        loading_time = abs(data['demands'][node]) * data['time_per_demand_unit']
        rest_time = 0
        if node in data['unload_node_list']:
            rest_time = 60 # suppose your problem is in minute
        return loading_time + rest_time

```

**Reply by orangiee** [reply]
Thank you so much for the fast reply and your answer's very helping.

**Comment by orangiee** [comment]
Hi, 
I have another questions about this code. How can I demande each driver to go to the reload depot and each driver have similar number of node to delivery rather than driver A have 14 nodes to delivery and driver B has no node. or driver A have no nodes in the morning and 14 nodes in the afternoon. 
1, for solving the problem of : demande each driver to go to the reload depot; I tried to add a very huge number to not drop reload node:

```python
for node in [1, 2, 3, 4, 5]:
        node_index = manager.NodeToIndex(node)
        routing.AddDisjunction([node_index], 100_000)
```

But it's not working. no car or just one car come to the reload depot. so how could I have each driver go to the reload depot. 

2, for solving the problem: make each driver have similar nodes to delivery, for example, driver A has 9 nodes in the morning, 10 in the afternoon; driver B has 8 in the morning and 12 in the afternoon; And all drivers do have nodes both in the morning and in the afternoon. So far, I tried to set max distance to a lower number so that to avoid one drive have a huge distance for his route, and other drivers have a small distances;  

```python
data['vehicle_max_distance'] = 160
```

but this seems not the solution. still, I got some drivers who have no node in the morning. Could you help me to solve this problem ? Thank you so much in advance !! 

3, I wonder can we set a value like : `data['vehicle_min_distance']` to limit a drive do at least 8 nodes in a day rather than 0 node. 

4, when I set `data['vehicle_max_time'] = 60000`, rather than `70200`, I got 'CP failed'; Could you explain this ? Thank you so much !

**Reply by Mizux** [reply]
1) You should assign each reload node to one vehicle using `routing.VehicleVar(index).SetValues([-1, vehicle_index])`

2/3) You can use `RoutingDimension::CumulVarSoftLowerBound()` with a counter dimension
https://github.com/google/or-tools/blob/86d4c543f717fa8716a9d66c25143660668bf825/ortools/constraint_solver/routing.h#L2666-L2676

4) do you have some TW above `60_000` ? i.e. if you limit vehicle capacity to K all your TW must be inside `[0, K]` otherwise your model is ill formed and solver return immediately at `solve()`

**Reply by orangiee** [reply]
Hi, 
Thank you for the reply. 
1. I tried your solution by adding `routing.VehicleVar(index).SetValues([-1, vehicle_index])` in the function of add_capacity_constraints or in the main part, but there still are one or two vehicles who would not go to the reload/unload depot. 
for me , I have 5 reloads nodes which correspond to 5  vehicles; for exempla, vehicle 0 go to reload node 1, vehicle 1 go to reload node 2, etc; I do like this: 

> for node in [1, 2, 3, 4, 5]:
>         node_index = manager.NodeToIndex(node)
>         routing.VehicleVar(node_index).SetValues([-1,node-1])
>         routing.AddDisjunction([node_index], 100000000000)

but it's not working;  Is there a way to force each vehicle go to reload node ? 

2/3. I tried add `SetCumulVarSoftLowerBound` in the part of `add_distance_dimension`:

> distance_dimension.SetCumulVarSoftLowerBound(
>                             distance_evaluator_index,
>                             60,
>                             100)

because I wanted to add an upper bound, so I added also: 

> distance_dimension.SetCumulVarSoftUpperBound(
>                             distance_evaluator_index,
>                             170,
>                             100)

But it seems that I still have some vehicles do just less than 60 meters or more than 170 meters. 

I also tried: 

> for vehicle_id in range(data['num_vehicles']):
>         ind = routing.End(vehicle_id)
>         distance_dimension.CumulVar(ind).RemoveInterval(0,60)
>         distance_dimension.CumulVar(ind).SetRange(60,170)

Could you please tell me which is right ? Thanks a lot. 

4. new question : how the `SetGlobalSpanCostCoefficient` influence the solver ? what the difference if I give distance_dimension.SetGlobalSpanCostCoefficient(10) and distance_dimension.SetGlobalSpanCostCoefficient(10000). 
I know how to calculate and it means to minimize the distances of routes , but I cannot understand how to set this value properly. Could you explain this ? thanks so much!

**Comment by orangiee** [comment]
Hi, should I create a new discussion ? please .
